### PR TITLE
Catch RuntimeError during handshake

### PR DESCRIPTION
**Pull request by cburgdorf**
### What was wrong?

A runtime error inside the event loop may happen during handshake and cause the node to crash (#590)

### How was it fixed?

Catch the `RuntimeError` and raise a `HanshakeFailure`

- [ ] add release notes entry

[//]: # (For important changes, please add a new entry to the running release notes PR)
[//]: # (You can find the current one using: https://github.com/ethereum/trinity/pulls?q=is%3Aopen+is%3Apr+label%3A%22Release+Notes%22 )
[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by pipermerriam** [comment]
Any chance you have the stacktrace laying around for this?

File: p2p/auth.py [hunk 1]
```diff
@@ -85,7 +85,11 @@
     initiator_nonce = keccak(os.urandom(HASH_LEN))
     auth_msg = initiator.create_auth_message(initiator_nonce)
     auth_init = initiator.encrypt_auth_message(auth_msg)
-    writer.write(auth_init)
+
+    try:
+        writer.write(auth_init)
+    except RuntimeError:
+        raise HandshakeFailure(f"Runtime error during handshake with {initiator.remote!r}")
```
Comment (pipermerriam) [inline p2p/auth.py]: Maybe this should be `raise HandshakeFailure("Error during handshake with {initiator.remote!r}: {err}") from err` to re-raise from original location of `RuntimeError`?

**Comment by carver** [comment]
Hm, I don't love catching such a generic exception and assuming we know why it was raised. At least we keep the underlying message in the `HandshakeFailure`, which is good. The only other option I can think of is to check if the writer transport is open before trying to write. I think that's better, but not enough to go make sure that it actually handles all the cases that are caught by the `RuntimeError` :) ... so I guess nothing to see here :see_no_evil:

**Comment by pipermerriam** [comment]
How about we check that the writer is closed in the exception block to try and *verify* that we're in the expected error state?  If somehow the write isn't closed we could just re-raise the original exception.

**Comment by cburgdorf** [comment]
@carver 

>Hm, I don't love catching such a generic exception and assuming we know why it was raised

Yes, I feel the same but couldn't think of a better way to fix it (and a bit of searching around revealed [aio-libs](https://github.com/aio-libs/) where applying the same strategy even though I seem unable to look it up again). 

> The only other option I can think of is to check if the writer transport is open before trying to write. I think that's better

Yes, maybe that's better.

>but not enough to go make sure that it actually handles all the cases that are caught by the `RuntimeError` :)

Right, but that's probably a good thing. So, if we happen to see another `RuntimeError` that is not related to the writer being closed, we can decide again on a case by case basis what to do.

@pipermerriam 

>How about we check that the writer is closed in the exception block to try and verify that we're in the expected error state? If somehow the write isn't closed we could just re-raise the original exception.

What's the main benefit of that? Not having to check `writer.transport.is_closing()` ([writer.is_closing()](https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamWriter.is_closing) is not available in Python 3.6) for the happy path? I don't think this is a performance critical path so not much to win here by avoiding that check imho.

So, I think, if we don't want to catch all `RuntimeError` at once, which I now feel convinced we shouldn't (thanks @carver for bringing it up), I think I would prefer to check `is_closing` upfront and not catch any `RuntimeError` at all and just see where that would take us :)

**Comment by pipermerriam** [comment]
@cburgdorf you can ignore my comment.  If there is a *pre-check* we can do instead of an *after the error occured* check, that's better.  I was just hastily reviewing this thread.

**Comment by cburgdorf** [comment]
Alright, send a PR to correct this https://github.com/ethereum/trinity/pull/614

**Pull request merged**